### PR TITLE
[SPARK-15108] [SQL] Describe Permanent UDTF

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

When Describe a UDTF, the command returns a wrong result. The command is unable to find the function, which has been created and cataloged in the catalog but not in the functionRegistry.

This PR is to correct it. If the function is not in the functionRegistry, we will check the catalog for collecting the information of the UDTF function. 
#### How was this patch tested?

Added test cases to verify the results

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @andrewor14 @yhuai @viirya

**Comment by viirya** [comment]
When we can't find the function in functionRegistry, does it mean it is not loaded yet?

**Comment by gatorsmile** [comment]
: ) Partially right. Even if the function is loaded, we might not be able to find it. The reason is we might not specify the database name in `Describe Function`. 

The current implementation of `FunctionRegistry` is not perfect. It misses a critical part. We need to introduce the concept of `database` there. I plan to submit another PR later for enhancing it.

**Comment by viirya** [comment]
yea. But I remember in Hive you can't find the function when you don't specify database name.

**Comment by viirya** [comment]
Does `FunctionRegistry` need to consider database? I think it is just a registry for functions in current database?

**Comment by viirya** [comment]
Besides, when a function is not loaded, I remember in Hive you can't find it with describe command too.

**Comment by gatorsmile** [comment]
First, we do not refresh `FunctionRegistry` when we switch the database. Thus, the current `FunctionRegistry` is a global registry. `FunctionRegistry` need a concept of database.

Second, when the function is loaded, `FunctionRegistry` has an entry `dbName.funcName`. This is not a clean way. 

Third, I need to confirm what is the behavior of Hive. If what you said is right, this looks strange to me. From users' perspective, they create a function but they are unable to describe it to verify if the function exists or not. They have to call the function for loading the function. This looks strange to me. Some function call could be very expensive. This design does not make sense to me. 

Forth, I think the users are allowed to describe a function without specifying the database name. At least, the parser allows it. If necessary, we can add the current database name when checking `FunctionRegistry`. So far, this is missing

Thanks for your inputs!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala [hunk 1]
```diff
@@ -129,7 +129,16 @@ case class DescribeFunction(
...
+        case None =>
+          val currentDb = sparkSession.sessionState.catalog.getCurrentDatabase
+          if (sparkSession.externalCatalog.functionExists(currentDb, functionName)) {
+            val catalogFunction = sparkSession.externalCatalog.getFunction(currentDb, functionName)
+            Row(s"Function: ${catalogFunction.identifier}") ::
+              Row(s"Class: ${catalogFunction.className}") ::
+              Row(s"Usage: N/A") :: Nil
+          } else {
+            Seq(Row(s"Function: $functionName not found."))
+          }
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: Looks like we only need to change https://github.com/apache/spark/pull/12885/files#diff-d2a203f08c862bd762e6740c16e972f7R118 to `sparkSession.sessionState.catalog.lookupFunctionInfo`?

**Comment by viirya** [comment]
Do we need to refresh `FunctionRegistry` when we switch the database? I think `FunctionRegistry` is like a registry for functions in current session. At least in Hive the function registration works like that.

**Comment by yhuai** [comment]
I do not think we need to do anything when we switch the database.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: : ) Yeah. We can do that in your way.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: We also need to change the parser and change the interface of `DescribeFunction` to:
```scala
case class DescribeFunction(
    functionName: FunctionIdentifier,
    isExtended: Boolean) extends RunnableCommand {
```

**Comment by gatorsmile** [comment]
cc @yhuai Let me know if you have any other suggestion. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala [hunk 2]
```diff
@@ -115,22 +116,23 @@ case class DescribeFunction(
         Row(s"Function: case") ::
           Row(s"Usage: CASE a WHEN b THEN c [WHEN d THEN e]* [ELSE f] END - " +
             s"When a = b, returns c; when a = d, return e; else return f") :: Nil
-      case _ => sparkSession.sessionState.functionRegistry.lookupFunction(functionName) match {
-        case Some(info) =>
+      case _ =>
+        try {
+          val info = sparkSession.sessionState.catalog.lookupFunctionInfo(functionName)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: Looks like this line is the only change that we need?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: nvm. Other changes look also needed. Let me take another look.

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 2.0.

**Comment by yhuai** [comment]
Thanks!

**Comment by gatorsmile** [comment]
Thank you!

**Pull request closed without merging**